### ISSUE TITLE
Add a way to map RecordBatch schema from one to another

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#5996]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
"Schema Evolution" is common when working with data -- this typically means that a table's overall schema may have columns that do not appear in individual `RecordBatch` 

For example, maybe initially data was written with (`user`, `address`) and then later on a new column is added (`user`, `address`, `user_comment`). 

All previously written data (e.g. in parquet files) might only have (`user`, `address`) 

| user | address |
|--------|--------|
| Andrew | USA |
| Marco | Germany | 

but new data has all three columns (`user`, `address`, `user_comment`).

| user | address | user_comment |
|--------|--------|--------|
| Andy | USA | Creator of DataFusion | 

To query this data as a single table typically requires the data to have the same schema, and what  query engines like Apache DataFusion or Apache Iceberg typically do is expand the old batches with null values, resulting in

| user | address | user_comment |
|--------|--------|--------|
| Andrew | USA | `NULL` |
| Marco | Germany | `NULL` |
| Andy | USA | Creator of DataFusion | 

**Describe the solution you'd like**
I would like
1.  an API in arrow-rs to do this mapping
2. Documentation and examples showing how to use it

**Describe alternatives you've considered**

#### SchemaMapper
One alternative would be to bring the `SchemaMapper` from DataFusion upstream to arrow-rs. This implementation has the benefit that it calculates the mapping once and then applies it to each batch

https://github.com/apache/datafusion/blob/b76c1b70506e45880f8e4bc5f9dc2ef088053acb/datafusion/core/src/datasource/schema_adapter.rs#L105-L174

#### `RecordBatch::ensure_schema`

Another alternative would be to add a `RecordBatch::ensure_schema` or something modeled on: 

https://github.com/influxdata/influxdb3_core/blob/0f5ecbd6b17f83f7ad4ba55699fc2cd3e151cf94/arrow_util/src/util.rs#L28-L47

where the API is smaller but it is less efficient

#### Leave / make public in DataFusion (rather than arrow-rs)

Another alternative might be to leave this API in DataFurstion (and make it public) rather than upstreaming it to arrow-rs

I am proposing adding it to arrow-rs as I think it is generally useful (we use it in InfluxDB several places, for example)

**Additional context**

**Comment by ion-elgreco** [comment]
It would be nice if that can be simplified a bit, I am overhauling our schema evolution handling in delta-rs at the moment

**Comment by alamb** [comment]
Adding a nice API in arrow-rs for this would be really valuable, in my opinion. Maybe we can reuse/ adapt what you have in delta-rs somehow

**Comment by ion-elgreco** [comment]
> Adding a nice API in arrow-rs for this would be really valuable, in my opinion. Maybe we can reuse/ adapt what you have in delta-rs somehow

Yeah I can take a look at it, maybe make my first PR for arrow-rs😜

**Comment by huymq1710** [comment]
take
